content/library/advanced-features/app-testing/get-started.md

```diff
@@ -0,0 +1,267 @@
+
+You can write as many tests (functions) within a single test script as you want. When calling `pytest` in a directory, all `test_<name>.py` files within it will be used for testing. This includes files within subdirectories. Each `test_<something>` function within those files will be executed as a test. You can place test files anywhere in your project directory, but it is common to collect tests into a designated `tests/` directory. For other ways to structure and execute tests, check out [How to invoke pytest](https://docs.pytest.org/how-to/usage.html) in the `pytest` docs.
+
+### Example project with app testing
+
+Consider the following project:
+
+```none
+myproject/
+├── app.py
+└── tests/
+    └── test_app.py
+```
+
+Main app file:
+
+```python
+"""app.py"""
+import streamlit as st
+
+# Initialize st.session_state.beans
+beans = st.session_state.get("beans", 0)
+
+st.title("Bean counter :paw_prints:")
+
+addend = st.number_input("Beans to add", 0, 10)
+if st.button("Add"):
+    st.session_state.beans += addend
+st.markdown(f"Toe beans counted: {st.session_state.beans}")
+```
+
+Testing file:
+
+```python
+"""test_app.py"""
+from streamlit.testing.v1 import AppTest
+
+def test_increment_and_add():
+    """A user increments the number input, then clicks Add"""
+    at = AppTest.from_file("app.py").run()
+    at.number_input[0].increment().run()
+    at.button[0].click().run()
+    assert at.markdown[0].value == "Toe beans counted: 1"
+```
+
+The main app file (`app.py`) contains four elements when rendered: `st.title`, `st.number_input`, `st.button`, and `st.markdown`. The test script (`test_app.py`) contains a single test (the function named `test_increment_and_add`). We'll cover how tests are written in more detail later on the page, but here's a brief explanation of what this test does:
+
+1. Initialize the simulated app and execute the first script run.
+   ```python
+   at = AppTest.from_file("app.py").run()
+   ```
+2. Simulate a user clicking the plus icon (<i style={{ verticalAlign: "-.25em" }} className={{ class: "material-icons-sharp" }}>add</i>) to increment the number input (and the resulting script rerun).
+   ```python
+   at.number_input[0].increment().run()
+   ```
+3. Simulate a user clicking the "**Add**" button (and the resulting script rerun).
+   ```python
+   at.button[0].click().run()
+   ```
+4. Check if the correct message is displayed at the end.
+   ```python
+   assert at.markdown[0].value == "Toe beans counted: 1"
+   ```
+
+Assertions are the heart of tests. When the assertion is true, the test passes. When the assertion is false, the test fails. A test can have multiple assertions, but it is a good practice to keep tests tightly focused. When tests focus on a single behavior, it is easier to understand and respond to failure.
+
+### Try out a simple test with `pytest`
+
+1. Install `pytest` into your Streamlit development environment if you haven't already done so:
+   ```bash
+   pip install pytest
```

Comment:
do we need to tell them to install streamlit too? Or is that safely assumed? (Whatever is the normal practice from other guides is fine, I just wasn't sure)

Reply:
I waffled on this, too. I do have it at the top as a "before you begin" but I was just a tad worried about someone jumping to the middle of the page from a good search and maybe missing that detail on the most rudimentary part. It is repetitious though, so I can take it out.